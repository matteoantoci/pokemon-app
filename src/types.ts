export enum PokemonType {
  Normal = 'Normal',
  Fire = 'Fire',
  Water = 'Water',
  Electric = 'Electric',
  Grass = 'Grass',
  Ice = 'Ice',
  Fighting = 'Fighting',
  Poison = 'Poison',
  Ground = 'Ground',
  Flying = 'Flying',
  Psychic = 'Psychic',
  Bug = 'Bug',
  Rock = 'Rock',
  Ghost = 'Ghost',
  Dragon = 'Dragon',
  Dark = 'Dark',
  Steel = 'Steel',
  Fairy = 'Fairy',
}

export type PokemonTypeColor = {
  [key in PokemonType]: string
}

export type Pokemon = {
  id: string
  image: string
  maxCP: number
  maxHP: number
  name: string
  number: string
  types: PokemonType[]
}
