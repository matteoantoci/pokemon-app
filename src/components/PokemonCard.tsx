import styled from 'styled-components'
import Typography from '@material-ui/core/Typography'
import { CardMedia, Grid, LinearProgress } from '@material-ui/core'
import Card from '@material-ui/core/Card'
import React, { useCallback } from 'react'
import CardContent from '@material-ui/core/CardContent'
import CardActions from '@material-ui/core/CardActions'
import Button from '@material-ui/core/Button'
import Tooltip from '@material-ui/core/Tooltip'
import { Flex } from './Flex'
import { Pokemon, PokemonType, PokemonTypeColor } from '../types'

const PokemonCardContents = styled(Card)`
  height: 100%;
  display: flex;
  flex-direction: column;
`

const PokemonImg = styled(CardMedia)`
  padding-top: 100%;
  margin: 8px;
  background-size: contain !important;
`

const InfoWrapper = styled(Flex)`
  display: inline-flex;
  flex-grow: 1;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin-bottom: 16px;
  &:last-child {
    margin-bottom: 0;
  }
`

const Name = styled(Typography)`
  flex: 1 1 auto;
`

const Number = styled(Typography)`
  flex: 0 0 auto;
`

const Progress = styled(LinearProgress)`
  width: 100%;
  height: 8px !important;
  margin-left: 16px;
`

const BadgeContainer = styled(Flex)`
  width: 100%;
  flex-direction: row;
  justify-content: flex-end;
`

const TYPE_COLORS: PokemonTypeColor = {
  [PokemonType.Normal]: '#A8A77A',
  [PokemonType.Fire]: '#EE8130',
  [PokemonType.Water]: '#6390F0',
  [PokemonType.Electric]: '#F7D02C',
  [PokemonType.Grass]: '#7AC74C',
  [PokemonType.Ice]: '#96D9D6',
  [PokemonType.Fighting]: '#C22E28',
  [PokemonType.Poison]: '#A33EA1',
  [PokemonType.Ground]: '#E2BF65',
  [PokemonType.Flying]: '#A98FF3',
  [PokemonType.Psychic]: '#F95587',
  [PokemonType.Bug]: '#A6B91A',
  [PokemonType.Rock]: '#B6A136',
  [PokemonType.Ghost]: '#735797',
  [PokemonType.Dragon]: '#6F35FC',
  [PokemonType.Dark]: '#705746',
  [PokemonType.Steel]: '#B7B7CE',
  [PokemonType.Fairy]: '#D685AD',
}

type TypeBadgeProps = {
  type: PokemonType
}

const TypeBadge: React.FC<TypeBadgeProps> = styled(Flex)`
  border-radius: 4px;
  background-color: ${({ type }: TypeBadgeProps) => TYPE_COLORS[type]};
  padding: 4px 6px;
  color: white;
  margin-left: 4px;
`

const MAX_SCALE = 3000

const getPercentage = (value: number) => Math.round((value * 100) / MAX_SCALE)

type PokemonCardProps = {
  pokemon: Pokemon
  onPokemonDetailsClick?: (id: string) => void
}

export const PokemonCard: React.FC<PokemonCardProps> = ({ pokemon, onPokemonDetailsClick, ...props }) => {
  const handleShowDetails = useCallback(() => {
    if (!onPokemonDetailsClick) return
    onPokemonDetailsClick(pokemon.id)
  }, [pokemon, onPokemonDetailsClick])
  return (
    <Grid {...props} item xs={12} sm={6} md={4}>
      <PokemonCardContents>
        <PokemonImg image={pokemon.image} title={pokemon.name} />
        <CardContent>
          <InfoWrapper>
            <Name variant="h5" component="h2">
              {pokemon.name}
            </Name>
            <Number color="textSecondary" variant="h5" component="p">
              {`#${pokemon.number}`}
            </Number>
          </InfoWrapper>
          <InfoWrapper>
            <Flex>ID:</Flex>
            <Typography variant="body2" color="textSecondary">
              {pokemon.id}
            </Typography>
          </InfoWrapper>
          <InfoWrapper>
            <Flex>Type:</Flex>
            <BadgeContainer>
              {pokemon.types.map(type => (
                <TypeBadge key={type} type={type}>
                  {type}
                </TypeBadge>
              ))}
            </BadgeContainer>
          </InfoWrapper>
          <InfoWrapper>
            <Flex>HP:</Flex>
            <Tooltip title={`Max HP: ${pokemon.maxHP}`}>
              <Progress variant="determinate" value={getPercentage(pokemon.maxHP)} />
            </Tooltip>
          </InfoWrapper>
          <InfoWrapper>
            <Flex>CP:</Flex>
            <Tooltip title={`Max CP: ${pokemon.maxCP}`}>
              <Progress variant="determinate" value={getPercentage(pokemon.maxCP)} />
            </Tooltip>
          </InfoWrapper>
        </CardContent>
        {onPokemonDetailsClick && (
          <CardActions>
            <Button size="small" color="primary" onClick={handleShowDetails}>
              More information
            </Button>
          </CardActions>
        )}
      </PokemonCardContents>
    </Grid>
  )
}
