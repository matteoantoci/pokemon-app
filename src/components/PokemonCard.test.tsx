import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { PokemonCard } from './PokemonCard'
import { Pokemon } from '../types'
import { getMockPokemon } from '../test.utils'

describe('Pokemon Card', () => {
  const mockPokemon: Pokemon = getMockPokemon('id')

  it('renders correctly', () => {
    const component = renderer.create(<PokemonCard pokemon={mockPokemon} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('when details click callback is used', () => {
    it('renders correctly', () => {
      const onPokemonDetailsClick = jest.fn()
      const component = renderer.create(
        <PokemonCard pokemon={mockPokemon} onPokemonDetailsClick={onPokemonDetailsClick} />
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('triggers callback', () => {
      const onPokemonDetailsClick = jest.fn()
      const wrapper = shallow(<PokemonCard pokemon={mockPokemon} onPokemonDetailsClick={onPokemonDetailsClick} />)

      const ctaButton = wrapper.find('[data-test="more-info-btn"]')
      ctaButton.simulate('click')

      expect(onPokemonDetailsClick).toHaveBeenCalledWith(mockPokemon.id)
    })
  })
})
