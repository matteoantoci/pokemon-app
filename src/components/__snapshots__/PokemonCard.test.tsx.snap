// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Card renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root sc-bwzfXH gPmFGo MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root sc-htpNat bxEaqd"
      style={
        Object {
          "backgroundImage": "url(\\"https://someimage.com\\")",
        }
      }
      title="name"
    />
    <div
      className="MuiCardContent-root"
    >
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <h2
          className="MuiTypography-root sc-ifAKCX bUzTRP MuiTypography-h5"
        >
          name
        </h2>
        <p
          className="MuiTypography-root sc-EHOje jPiFzx MuiTypography-h5 MuiTypography-colorTextSecondary"
        >
          #number
        </p>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          ID:
        </div>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          id
        </p>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          Type:
        </div>
        <div
          className="sc-gzVnrw iKWuGe sc-bdVaJa kZAgjJ"
        >
          <div
            className="sc-htoDjs dCzLKI sc-bdVaJa kZAgjJ"
            type="Ghost"
          >
            Ghost
          </div>
          <div
            className="sc-htoDjs gbdiEo sc-bdVaJa kZAgjJ"
            type="Normal"
          >
            Normal
          </div>
        </div>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          HP:
        </div>
        <div
          aria-describedby={null}
          aria-valuenow={15}
          className="MuiLinearProgress-root MuiLinearProgress-colorPrimary sc-bZQynM hzDXFo MuiLinearProgress-determinate"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="progressbar"
          title="Max HP: 456"
        >
          <div
            className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
            style={
              Object {
                "transform": "translateX(-85%)",
              }
            }
          />
        </div>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          CP:
        </div>
        <div
          aria-describedby={null}
          aria-valuenow={4}
          className="MuiLinearProgress-root MuiLinearProgress-colorPrimary sc-bZQynM hzDXFo MuiLinearProgress-determinate"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="progressbar"
          title="Max CP: 123"
        >
          <div
            className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
            style={
              Object {
                "transform": "translateX(-96%)",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokemon Card when details click callback is used renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root sc-bwzfXH gPmFGo MuiPaper-rounded"
  >
    <div
      className="MuiCardMedia-root sc-htpNat bxEaqd"
      style={
        Object {
          "backgroundImage": "url(\\"https://someimage.com\\")",
        }
      }
      title="name"
    />
    <div
      className="MuiCardContent-root"
    >
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <h2
          className="MuiTypography-root sc-ifAKCX bUzTRP MuiTypography-h5"
        >
          name
        </h2>
        <p
          className="MuiTypography-root sc-EHOje jPiFzx MuiTypography-h5 MuiTypography-colorTextSecondary"
        >
          #number
        </p>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          ID:
        </div>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          id
        </p>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          Type:
        </div>
        <div
          className="sc-gzVnrw iKWuGe sc-bdVaJa kZAgjJ"
        >
          <div
            className="sc-htoDjs dCzLKI sc-bdVaJa kZAgjJ"
            type="Ghost"
          >
            Ghost
          </div>
          <div
            className="sc-htoDjs gbdiEo sc-bdVaJa kZAgjJ"
            type="Normal"
          >
            Normal
          </div>
        </div>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          HP:
        </div>
        <div
          aria-describedby={null}
          aria-valuenow={15}
          className="MuiLinearProgress-root MuiLinearProgress-colorPrimary sc-bZQynM hzDXFo MuiLinearProgress-determinate"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="progressbar"
          title="Max HP: 456"
        >
          <div
            className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
            style={
              Object {
                "transform": "translateX(-85%)",
              }
            }
          />
        </div>
      </div>
      <div
        className="sc-bxivhb iXtRbx sc-bdVaJa kZAgjJ"
      >
        <div
          className="sc-bdVaJa kZAgjJ"
        >
          CP:
        </div>
        <div
          aria-describedby={null}
          aria-valuenow={4}
          className="MuiLinearProgress-root MuiLinearProgress-colorPrimary sc-bZQynM hzDXFo MuiLinearProgress-determinate"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="progressbar"
          title="Max CP: 123"
        >
          <div
            className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
            style={
              Object {
                "transform": "translateX(-96%)",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall"
        data-test="more-info-btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          More information
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
